common.xml: add smart_battery_info message

hendjoshsr71 opened the pull request:
Add smart_battery_info message.

One small documentation change on the MAVLink side here: https://github.com/mavlink/mavlink/pull/1551

peterbarker commented:
@WickedShell

tridge commented:
we'll let @WickedShell have a look at this

tridge commented:
looks good to me

tridge commented on message_definitions/v1.0/common.xml:
```diff
@@ -4935,6 +4935,24 @@
+      <field type="char[16]" name="serial_number">Serial number in ASCII characters, 0 terminated. All 0: field not provided.</field>
+      <field type="char[50]" name="device_name">Static device name. Encode as manufacturer and product names separated using an underscore.</field>
+      <field type="uint16_t" name="weight" units="g">Battery weight. 0: field not provided.</field>
+      <field type="uint16_t" name="discharge_minimum_voltage" units="mV">Minimum per-cell voltage when discharging. If not supplied set to UINT16_MAX value.</field>
```
if this is per-cell, then where do we say the number of cells? I think we should have a uint8_t num_cells

hendjoshsr71 replied:
Makes sense to add that. The other item I thought about is discharge_maximum_current?

auturgy replied:
Remember that this message doesn't replace     <message id="147" name="BATTERY_STATUS"> but supplements it, so I think number of cells should be pulled out of the cell voltages sent in that message.  Discharge_maximum_current is potentially problematic, as it varies with voltage, and smart batteries should be managing this anyway.  I agree it would be useful in some circumstances though.

hendjoshsr71 replied:
Discharge_current_maximum doesn't vary with voltage as it is a manufacturer specification.  Perhaps you meant discharge_maximum_power? Only reason I bring it up is that this message and  message 147 seems to capture most of the important info. regarding a pack except the max. discharge current. And I saw the use of a few parameters to limit motor power based upon maximum discharge.

WickedShell replied:
Notably nothing in the message seems to tell you what maximum voltage is...

auturgy replied:
I think the intent was that max voltage is inferred from chemistry and number of cells, but valid point.
We’re starting to scope creep though.

hendjoshsr71 replied:
I'm wondering if the charging_minimum_voltage was really meant to be charging_maximum_voltage? It is a little odd for an autopilot to care about that as usually is it only for chargers to prevent charging of a deeply discharged battery.

peterbarker replied:
@hendjoshsr71 on the min-voltage thing - hybrid vehicles are a thing...

hendjoshsr71 replied:
Added these items

peterbarker commented:
Are we going to suggest adding that maximum-voltage thing upstream?

hendjoshsr71 commented:
Organizing common/important battery specs not in this message currently for discussion.
- `charging_maximum_voltage`
- `discharge_maximum_current`
- `discharge_burst_current`   (somewhat covered by previous)
- `manufacture_date` (could concatenate this into product name? From looking through specs and TI ICs `manufacturer_name` and `device_name` are a max. of 12 chars each )

hendjoshsr71 commented:
Added all of the items above to the message and created an upstream PR https://github.com/mavlink/mavlink/pull/1618

hamishwillee commented:
FYI, I reviewed in common - https://github.com/mavlink/mavlink/pull/1618#issuecomment-824547417 . My main concern is whether there are any known implementations that will be broken by this (yes it is marked WIP, but as you all know, we're trying to have common only for truly solid stuff. So if this has no implementations it really should be in development.xml.

hendjoshsr71 commented:
Rebased and changed the name of `num_cells` to `cells_in_series` for clarity 

@hamishwillee I have a PR in progress to use this message here https://github.com/ArduPilot/ardupilot/pull/16167.

hamishwillee commented:
> Rebased and changed the name of num_cells to cells_in_series for clarity

@hendjoshsr71 Sounds like a good change. Note I am assuming that you're pushing these upstream into mavlink/mavlink since that is what get's signed off in the repo.

hendjoshsr71 commented:
Just pushed these changes to mavlink/mavlink. I will update the ArduPilot side with the manufacture data format that comes out from the Mavlink Dev call.

hendjoshsr71 commented:
Made these new fields extensions, removed WIP tags, and switched to `Maximum per-cell voltage when charged` using an uint16_t

hamishwillee commented:
@peterbarker The upstream version has some minor differences to this, but I think we're pretty happy with it. Can you please confirm it also makes sense to you/"the ardupilot community" before I merge. Otherwise I'll wait for James to review in devcall next week.

hendjoshsr71 commented:
I haven't updated this version yet because I was waiting to get the upstream side settled first. :)

hamishwillee commented:
@hendjoshsr71 That's fair enough except that what will ideally happen is that someone will pull your changes down from mavlink/mavlink rather than merge this. But we still want the message to make sense to everyone - so I'm hoping Peter can check the mavlink version.

hamishwillee commented:
Well, it got merged, so you can either sync to this or wait for it to come from upstream.
